accord-core/src/main/java/accord/local/Command.java

```diff
@@ -399,158 +525,253 @@ private void onChangeInternal(PartialCommand command)
+        notifyListeners(safeStore);
     }
 
-    private static Function<CommandStore, Void> callPostApply(TxnId txnId)
+    private static Function<SafeCommandStore, Void> callPostApply(TxnId txnId)
     {
-        return commandStore -> {
-            commandStore.command(txnId).postApply();
+        return safeStore -> {
+            safeStore.command(txnId).postApply(safeStore);
             return null;
         };
     }
 
-    protected Future<Void> apply()
+    protected Future<Void> apply(SafeCommandStore safeStore)
     {
         // important: we can't include a reference to *this* in the lambda, since the C* implementation may evict
         // the command instance from memory between now and the write completing (and post apply being called)
-        return writes().apply(commandStore()).flatMap(unused ->
-            commandStore().process(this, callPostApply(txnId()))
+        CommandStore unsafeStore = safeStore.commandStore();
+        return writes().apply(safeStore).flatMap(unused ->
+            unsafeStore.submit(this, callPostApply(txnId()))
         );
     }
 
-    public Future<Data> read(Keys scope)
+    public Future<Data> read(SafeCommandStore safeStore)
     {
-        return txn().read(this);
+        return partialTxn().read(safeStore, this);
     }
 
-    private Future<Void> maybeExecute(boolean notifyListenersOnNoop)
+    // TODO: maybe split into maybeExecute and maybeApply?
+    private boolean maybeExecute(SafeCommandStore safeStore, ProgressShard shard, boolean alwaysNotifyListeners, boolean notifyWaitingOn)
     {
         if (logger.isTraceEnabled())
-            logger.trace("{}: Maybe executing with status {}. Will notify listeners on noop: {}", txnId(), status(), notifyListenersOnNoop);
+            logger.trace("{}: Maybe executing with status {}. Will notify listeners on noop: {}", txnId(), status(), alwaysNotifyListeners);
 
-        if (status() != Committed && status() != Executed)
+        if (status() != Committed && status() != PreApplied)
         {
-            if (notifyListenersOnNoop) notifyListeners();
-            return Writes.SUCCESS;
+            if (alwaysNotifyListeners)
+                notifyListeners(safeStore);
+            return false;
         }
 
-        if (isUnableToApply())
+        if (isUnableToExecute())
         {
-            BlockedBy blockedBy = blockedBy();
-            if (blockedBy != null)
-            {
-                logger.trace("{}: not executing, blocked on {}", txnId(), blockedBy.command.txnId());
-                commandStore().progressLog().waiting(blockedBy.command, blockedBy.someKeys);
-                if (notifyListenersOnNoop) notifyListeners();
-                return Writes.SUCCESS;
-            }
-            assert !isWaitingOnApply();
+            if (alwaysNotifyListeners)
+                notifyListeners(safeStore);
+
+            if (notifyWaitingOn)
+                new NotifyWaitingOn(this).accept(safeStore);
```

Comment:
We don't currently, we only prevent re-entry. If we are concerned about this, I'll add an API mechanism to at least minimise it. There's no issue having multiple, but it could be wasteful, so a best-effort mechanism will be fine.

Reply:
OK, I'll leave it up to you if you want to do it now. I'm not sure if this is actually worse than blockedBy from before.